lib/url-decorator.jsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { CompositeDecorator } from 'draft-js';
+import urlRegex from 'url-regex';
+import normalizeUrl from 'normalize-url';
+import { viewExternalUrl } from './utils/url-utils';
+
+const urlRe = urlRegex();
+/*
+- markdownLinkRe is a regexp to be applied to the substring preceding a matched url, ie. `wholeString.slice(0, wholeString.indexOf(matchedUrl))`
+- it tests whether the matched url is preceded by the charcters that define a link in markdown, ie. `(` or `]: `
+*/
+const markdownLinkRe = /(?:\(|\]:\s+)$/;
+
+export function normalizeUrlsInMarkdownString( str ) {
+	return str.replace( urlRe, ( match, offset ) => {
+		var normalized = normalizeUrl( match );
+		if ( markdownLinkRe.test( str.slice( 0, offset ) ) ) {
+			return normalized;
+		}
+		// replace url with markdown link only if it's not already part of one
+		return `[${match}](${normalized})`;
+	} );
+}
```

Comment:
> why isn't our Markdown parser detecting the URLs… shouldn't Markdown automatically linkify all the things?

Regular markdown only linkifies things that are defined with the link syntax, either `[abc](www.abc.com)` or `[abc][1]`. Github flavoured markdown is the one that linkifies everything.

> also, what about <a href="wordpresscom">Yahoo</a> - what happens here?

well in that particular case you get shown a link with an invalid href because that doens't match the url regexp. however, if that href was valid this would actually break :( and replace the href with a markdown url...
I'm starting to think it's actually not a great idea to be writing this function, how about using a different `markdown` parser (instead of `marked`) that actually linkifies everything (eg. https://github.com/markdown-it/markdown-it)?